src/pubsub/ua_pubsub_keystorage.h

```diff
@@ -0,0 +1,396 @@
+ * milli seconds
+ * @return UA_StatusCode the return status
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_storeSecurityKeys(UA_Server *server, UA_PubSubKeyStorage *keyStorage,
+                                      UA_UInt32 currentTokenId, const UA_ByteString *currentKey,
+                                      UA_ByteString *futureKeys, size_t futureKeyCount,
+                                      UA_Duration msKeyLifeTime);
+
+/**
+ * @brief Finds the KeyItem from the KeyList by KeyId
+ *
+ * @param keyId the identifier of the Key
+ * @param keyStorage pointer to the keystorage
+ * @param keyItem returned pointer to the keyItem in the KeyList
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_getKeyByKeyID(const UA_UInt32 keyId, UA_PubSubKeyStorage *keyStorage,
+                                  UA_PubSubKeyListItem **keyItem);
+
+/**
+ * @brief Adds a new KeyItem at the end of the KeyList
+ * to the new KeyListItem.
+ *
+ * @param keyStorage pointer to the keystorage
+ * @param key the key to be added
+ * @param keyID the keyID associated with the key to be added
+ */
+UA_PubSubKeyListItem *
+UA_PubSubKeyStorage_push(UA_PubSubKeyStorage *keyStorage, const UA_ByteString *key,
+                         UA_UInt32 keyID);
+
+/**
+ * @brief It calculates the time to trigger the callback to update current key, adds the
+ * callback to the server and returns the callbackId.
+ *
+ * @param server the server object
+ * @param keyStorage the pointer to the existing keystorage in the server
+ * @param callback the callback function to be added to the server
+ * @param timeToNextMs time in milli seconds to trigger the callback function
+ * @param callbackID the returned callbackId of the added callback function
+ * @return UA_StatusCode the return status
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_addMoveToNextKeyCallback(UA_Server *server,
+                                          UA_PubSubKeyStorage *keyStorage,
+                                          UA_ServerCallback callback,
+                                          UA_Duration timeToNextMs,
+                                          UA_UInt64 *callbackID);
+
+/**
+ * @brief It takes the current Key data, divide it into singing key, encrypting key and
+ * keyNonce according to security policy associated with PubSub Group and  set it in
+ * channel context of the assocaited PubSub Group. In case of pubSubGroupId is
+ * UA_NODEID_NULL, all the Reader/WriterGroup's channelcontext are updated with matching
+ * SecurityGroupId.
+ *
+ * @param server The server object
+ * @param pubSubGroupId the nodeId of the Reader/WirterGroup whose channel context to be
+ * updated
+ * @param securityGroupId The identifier for the SecurityGroup
+ * @return UA_StatusCode return status code
+ */
+UA_StatusCode
+UA_PubSubKeyStorage_importKeyToChannelContext(UA_Server *server, const UA_NodeId pubSubGroupId,
```

Comment:
This is used in only one place and does not have to be exposed.
It makes sense that the keys are loaded when the rollover happens (and only then).
Make the function `static`, i.e. defined only in the compilation unit.
Also, the only place that uses it calls with NODEID_NULL.

Reply:
There are technically three scenarios for this method here

1. When the SetSecurityKeys Method is called, it passes the current and future key. The current key is immediately imported into channel context with this method.
2. When keylifetime expires, the Rollover moves the currentItem pointer to the next key and then imports that key to the channelcontext. The first iteration happens after the first keylifetime period expires, which can take any amount of time.
3. When keys are fetched using Pull Model (GeSecurityKeys Method) and stored in the keyStorage and the currentItem points to the current key. Then Writer/Reader groups can be into operational at random times. Then the current key is imported from KeyStorage. It is also possible that the Rollover has already done multiple iterations before setting into Operational state, in this Channel Context will be update with same key material.

These are the reasons I set it non static.

NODEID_NULL, is used to set the keys for all the related Writer/Reader Groups. However, when a nodeId is given only that Reader/Writer Group is update. It is used in the next patches related to SKS client.

Reply:
Ok. So there is additional usage that will only come with the future PR.
I think the name _activateKey makes more sense.
Import sounds like importing from a file.

Reply:
changed it to _activateKeyToChannelContext()